File: Design-Patterns/Visitor/Cpp/remove_broken_armor_visitor.cpp
#include "equipment_visitor.h"
#include <iostream>

remove_broken_armor_visitor::remove_broken_armor_visitor(std::shared_ptr<backpack> backpack)
{
	backpack_ = backpack;
}

void remove_broken_armor_visitor::visit_armor(std::shared_ptr<armor> armor)
{
	if (armor->get_endurance() == 0)
	{
		std::cout << "Armor \"" + armor->get_name() + "\" is broken, remove it." << std::endl;
		backpack_->remove(armor);		
	}
}

File: Design-Patterns/Visitor/CSharp/Visitors/RemoveBrokenArmorVisitor.cs
using System;
using Visitor.Equipments;

namespace Visitor.Visitors
{
    public class RemoveBrokenArmorVisitor : EquipmentVisitor
    {
        private readonly Backpack _backpack;

        public RemoveBrokenArmorVisitor(Backpack backpack)
        {
            _backpack = backpack;
        }

        public override void VisitArmor(Armor armor)
        {
            if (armor.Endurance == 0)
            {
                _backpack.Remove(armor);
                Console.WriteLine($"Armor \"{armor.Name}\" is broken, remove it.");
            }
        }
    }
}
